src/swupd/fullfiles.go

```diff
@@ -0,0 +1,274 @@
+
+		var err error
+		switch f.Type {
+		case typeDirectory:
+			err = createDirectoryFullfile(input, name, output)
+		case typeLink:
+			err = createLinkFullfile(input, name, output)
+		case typeFile:
+			err = createRegularFullfile(input, name, output)
+		default:
+			return fmt.Errorf("cannot create fullfiles: file %s is of unsupported type %q", f.Name, f.Type)
+		}
+
+		if err != nil {
+			return err
+		}
+
+		done[f.Hash] = true
+	}
+	return nil
+}
+
+func createDirectoryFullfile(input, name, output string) error {
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if !fi.IsDir() {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a directory but it is not", input)
+	}
+
+	hdr, err := getHeaderFromFileInfo(fi)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	hdr.Name = name
+	hdr.Typeflag = tar.TypeDir
+
+	err = createTarGzipHeaderOnly(output, hdr)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	return nil
+}
+
+func createLinkFullfile(input, name, output string) error {
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if fi.Mode()&os.ModeSymlink == 0 {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a link but it is not", input)
+	}
+	link, err := os.Readlink(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	hdr, err := getHeaderFromFileInfo(fi)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	hdr.Name = name
+	hdr.Typeflag = tar.TypeSymlink
+	hdr.Linkname = link
+
+	err = createTarGzipHeaderOnly(output, hdr)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	return nil
+}
+
+func createRegularFullfile(input, name, output string) (err error) {
+	// Ensure this is a regular file.
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if !fi.Mode().IsRegular() {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a regular file but it is not", input)
+	}
+
+	// Create the uncompressed fullfile.
+	uncompressed, err := os.Create(output)
+	if err != nil {
+		return fmt.Errorf("couldn't create uncompressed fullfile: %s", err)
+	}
+	defer func() {
+		cerr := uncompressed.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	err = tarRegularFullfile(uncompressed, input, name, fi)
+	if err != nil {
+		return fmt.Errorf("couldn't archive the file %s: %s", input, err)
+	}
+	// Sync the data so a compressor (possibly external program) can read this file while we
+	// still have an open file to pass to the compressors that read directly from it.
+	err = uncompressed.Sync()
```

Comment:
We shouldn't really need this. I had in mind external tools reusing the tar we just wrote, but we don't have or plan to have that. I'll let a note around and if that ever happens we close/reopen or sync the file.